pysmt/parsing.py

```diff
@@ -0,0 +1,435 @@
+#
+# This file is part of pySMT.
+#
+#   Copyright 2014 Andrea Micheli and Marco Gario
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+
+import re
+from fractions import Fraction
+
+from pysmt.shortcuts import get_env
+
+
+##
+## Precedence table (from low priority to high priority):
+##
+# 10 ## Xor, Implies, Iff
+# 20 ## Forall, Exists
+# 30 ## Or
+# 40 ## And
+# 50 ## Not
+# 60 ## LE, LT, Equals, GE, GT, BVULE, BVUGE, BVULT, BVUGT, BVSLE, BVSGE, BVSGT
+# 70 ## Plus, Minus, BVAdd, BVSub
+# 80 ## Times, Div, BVMul, BVUDiv, BVSDiv, BVSRem, BVURem
+# 90 ## BVLShl, BVLShr, BVAShr, BVConcat, BVXor, BVRor, BVRol, BVZExt, BVSExt, BVComp
+# 100 ## ToReal Uminus BVNeg
+# 200 ## ()
+# 300 ## []
+
+
+class HRLexer(object):
+    """This class produces a stream of token objects for the Pratt-parser"""
+
+    def __init__(self, env=None):
+        if env is None:
+            env = get_env()
+        self.env = env
+        self.mgr = env.formula_manager
+        self.get_type = env.stc.get_type
+
+        self.rules = []
```

Comment:
Where are "rules" defined?
You are appending (a lot) of tuples, but the semantics of the tuple is unclear.